### ISSUE TITLE
Adding titles/labels to submitted links

### ISSUE THREAD

**Issue by xander-marjoram** [tommcclean/cmpct.io#27]
**Is your feature request related to a problem? Please describe.**
When creating a list of links, or visiting a landing page that contains a list of links, some of the links may be quite opaque, and it may not be immediately obvious what that page will contain. For example, a list of Google spreadsheets, or YouTube videos.

**Describe the solution you'd like**
On the landing page, I would like to be able to see some additional context about what the page contains, rather than having to scan the link for any clues.

**Describe solutions you've considered**
- A way of manually entering a heading/label for each link as the list is created, although this adds friction. This has the potential to be more concise and appropriate, but also to be misleading.
- Fetching the title of the page and displaying this.

I believe fetching the title would be the most effective solution, as page titles are usually descriptive enough to provide sufficient context, it does not require additional input when creating the link, and it may also provide some value for single links as well as lists. If the owner of the link is adhering to best practice, then this title should also be reasonably short which is an added benefit.

**Comment by tommcclean** [comment]
Interesting idea, I am investigating this with a proof of concept: https://github.com/tommcclean/functions.cmpct.io

The repo in that link is a new Azure Function Set, the RouteProcessor is triggered by a route JSON file being written to the storage container, on trigger it iterates over each link and queries the link securely to obtain the page title. It then uses that information to rewrite the route file, so that subsequent requests through the website will be able to get the title back.

I decided not to trial this in the API, as I didn't want to do anything long running or blocking in the API that wasn't needed for the core product to work, but as this is isolated from the front end and the API, it can gracefully increase the quality of data over time.

The proof of concept will run for a little while, to see how performant it is, how much it costs to run, and chiefly, how long does it take from a route being created, for it to be updated with associated metadata. If the reaction time is fast, and it is cheap to operate, I will extend it to also scrape for a screenshot of the page, which will start to exist in the data, and can be made visible on the front end website later.

With this prototype, the model is rehydrated as follows.
![image](https://user-images.githubusercontent.com/10741583/73695771-b4494a80-46d2-11ea-9cd9-cd5068b02826.png)

This solution won't work for links hidden behind authentication however. Places like slack tend to get around this by using the user's auth token with the service, but that is not achievable for this project.

This functionality can also be extended to automatically report dead links.

**Comment by tommcclean** [comment]
This Proof of Concept seems to be quite successful. So I have pointed it at the main data store, and it has processed the majority of the content quickly. The API has been updated to extend the response model to include a new title property.

Finally the front end has been updated to apply the title on hover.

When more information is gathered about a link, a modal or some other UI element could be used to display it, but as an early attempt; a title will suffice to see how it works out.

As demonstrated here - https://cmpct.io/mJyxBoI1

The delay in adding this data to the model takes approx a second, so as soon as it is shared; such content is available.

![image](https://user-images.githubusercontent.com/10741583/73775912-b8cb3d00-477e-11ea-89c0-6406458906f7.png)

**Comment by tommcclean** [comment]
Oh and it automatically reports dead links
![image](https://user-images.githubusercontent.com/10741583/73777547-92f36780-4781-11ea-8a81-84a3ec284c9f.png)

**Comment by tommcclean** [comment]
I have decided to stop auto reporting for now, as it reports some links that may be network blocked as opposed to anything else.

I have extended the front end to show page titles more prominently where they can be acquired. Where they can't; there is no UI change. I think this solves your feature request.

![image](https://user-images.githubusercontent.com/10741583/73886077-cfe25b80-4860-11ea-8cfc-1711a7ef4ac4.png)
